File: api/src/dressme/db/seed.py
from pathlib import Path
from typing import TypedDict
from uuid import uuid4

from sqlmodel import Session

from ..settings import get_settings
from ..blob_storage import get_blob_storage
from ..image_utils import get_content_type_from_path
from ..wearable_classification import WearableCategory
from . import create_db_and_tables, engine
from .models import User, Wearable, WearableOnAvatarImage


class WearableSeedData(TypedDict):
    name: str
    category: WearableCategory
    image_path: str

settings = get_settings()

selfie_data = {"name": "human_4", "image_path": "images/humans/selfie_4.jpg"}
avatar_data = {"name": "human_4", "image_path": "images/avatars/avatar_4.jpg"}

wearables_data: dict[str, WearableSeedData] = {
    "tshirt": {
        "name": "tshirt",
        "category": "t-shirt",
        "image_path": "images/wearables/tops/t-shirt/purple-tshirt-product.webp",
    },
    "shirt": {
        "name": "shirt",
        "category": "shirt",
        "image_path": "images/wearables/tops/shirt/button-down-casual.jpeg",
    },
    "sweater": {
        "name": "sweater",
        "category": "sweater",
        "image_path": "images/wearables/tops/sweater/pullover-casual.webp",
    },
    "jacket": {
        "name": "jacket",
        "category": "jacket",
        "image_path": "images/wearables/tops/jacket/blazer-casual.webp",
    },
    "top": {
        "name": "top",
        "category": "top",
        "image_path": "images/wearables/tops/top/basic-top-product.webp",
    },
    "pants": {
        "name": "pants",
        "category": "pants",
        "image_path": "images/wearables/bottoms/pants/jeans-product.webp",
    },
    "shorts": {
        "name": "shorts",
        "category": "shorts",
        "image_path": "images/wearables/bottoms/shorts/gym-shorts-product.webp",
    },
    "skirt": {
        "name": "skirt",
        "category": "skirt",
        "image_path": "images/wearables/bottoms/skirt/mini-skirt-product.webp",
    },
}

# Path to the repo root
ROOT_PATH = Path(__file__).parent.parent.parent.parent.parent


def seed():
    create_db_and_tables()
    blob_storage = get_blob_storage()

    with Session(engine) as session:
        # Upload selfie image
        selfie_image_path = ROOT_PATH / Path(selfie_data["image_path"])
        with open(selfie_image_path, "rb") as image_file:
            selfie_image_data = image_file.read()

        selfie_image_key = f"{uuid4()}.jpg"
        blob_storage.upload(
            settings.SELFIES_BUCKET,
            selfie_image_key,
            selfie_image_data,
            get_content_type_from_path(selfie_image_path),
        )

        # Upload avatar image
        avatar_image_path = ROOT_PATH / Path(avatar_data["image_path"])
        with open(avatar_image_path, "rb") as image_file:
            avatar_image_data = image_file.read()

        avatar_image_key = f"{uuid4()}.jpg"
        blob_storage.upload(
            settings.AVATARS_BUCKET,
            avatar_image_key,
            avatar_image_data,
            get_content_type_from_path(avatar_image_path),
        )

        # Add user
        if settings.AUTH0_SEED_USER_ID is None:
            raise ValueError(
                "AUTH0_SEED_USER_ID is not set, but this is required to determine which user should own the data added during seeding. You can find this your user ID in the Auth0 dashboard under User Management."
            )
        user = User(
            auth0_user_id=settings.AUTH0_SEED_USER_ID,
            selfie_image_key=selfie_image_key,
            avatar_image_key=avatar_image_key,
        )
        session.add(user)

        # Add wearables
        for wearable_data in wearables_data.values():
            # Upload wearable image
            wearable_image_path = ROOT_PATH / Path(wearable_data["image_path"])
            with open(wearable_image_path, "rb") as image_file:
                wearable_image_data = image_file.read()

            wearable_image_key = f"{uuid4()}{wearable_image_path.suffix}"
            blob_storage.upload(
                settings.WEARABLES_BUCKET,
                wearable_image_key,
                wearable_image_data,
                get_content_type_from_path(wearable_image_path),
            )

            # Add wearable
            wearable = Wearable(
                category=wearable_data["category"],
                image_key=wearable_image_key,
                user_id=user.id,
            )
            session.add(wearable)

            # Upload WOA image
            woa_image_path = (
                ROOT_PATH
                / "images"
                / "results"
                / avatar_data["name"]
                / "single"
                / f"{wearable_data['name']}.jpg"
            )
            with open(woa_image_path, "rb") as image_file:
                woa_image_data = image_file.read()

            woa_image_key = f"{uuid4()}.jpg"
            blob_storage.upload(
                settings.WOA_BUCKET,
                woa_image_key,
                woa_image_data,
                "image/jpeg",
            )

            # Upload mask image
            mask_image_path = (
                ROOT_PATH
                / "images"
                / "masks"
                / avatar_data["name"]
                / "post"
                / f"{wearable_data['name']}.jpg"
            )
            with open(mask_image_path, "rb") as mask_image_file:
                mask_image_data = mask_image_file.read()

            mask_image_key = f"{uuid4()}.jpg"
            blob_storage.upload(
                settings.WOA_BUCKET,
                mask_image_key,
                mask_image_data,
                "image/jpeg",
            )

            # Add WearableOnAvatarImage
            wearable_on_avatar_image = WearableOnAvatarImage(
                user_id=user.id,
                avatar_image_key=avatar_image_key,
                wearable_image_key=wearable_image_key,
                image_key=woa_image_key,
                mask_image_key=mask_image_key,
            )
            session.add(wearable_on_avatar_image)

        session.commit()


if __name__ == "__main__":
    seed()

File: scripts/seed.ts
import "dotenv/config";
import { eq } from "drizzle-orm";
import { lookup } from "mime-types";
import { randomUUID } from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { getDb, schema } from "../src/server/db";
import { getBlobStorage } from "../src/server/services";
import { getSettings } from "../src/server/settings";

const settings = getSettings();

// Path to the repo root
const ROOT_PATH = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");

const SELFIE_PATH = "images/humans/selfie_4.jpg";
const AVATAR_PATH = "images/avatars/avatar_4.jpg";

interface WearableSeedData {
  name: string;
  category: string;
  imagePath: string;
}

const WEARABLES: Record<string, WearableSeedData> = {
  tshirt: {
    name: "tshirt",
    category: "t-shirt",
    imagePath: "images/wearables/tops/t-shirt/purple-tshirt-product.webp",
  },
  shirt: {
    name: "shirt",
    category: "shirt",
    imagePath: "images/wearables/tops/shirt/button-down-casual.jpeg",
  },
  sweater: {
    name: "sweater",
    category: "sweater",
    imagePath: "images/wearables/tops/sweater/pullover-casual.webp",
  },
  jacket: {
    name: "jacket",
    category: "jacket",
    imagePath: "images/wearables/tops/jacket/blazer-casual.webp",
  },
  top: {
    name: "top",
    category: "top",
    imagePath: "images/wearables/tops/top/basic-top-product.webp",
  },
  pants: {
    name: "pants",
    category: "pants",
    imagePath: "images/wearables/bottoms/pants/jeans-product.webp",
  },
  shorts: {
    name: "shorts",
    category: "shorts",
    imagePath: "images/wearables/bottoms/shorts/gym-shorts-product.webp",
  },
  skirt: {
    name: "skirt",
    category: "skirt",
    imagePath: "images/wearables/bottoms/skirt/mini-skirt-product.webp",
  },
};

async function seed() {
  const seedUserId = settings.AUTH0_SEED_USER_ID;
  if (!seedUserId) {
    throw new Error(
      "AUTH0_SEED_USER_ID is not set, but this is required to determine which user " +
        "should own the data added during seeding.",
    );
  }

  const db = getDb();
  const blobStorage = getBlobStorage();

  // Upload selfie image
  const selfiePath = path.join(ROOT_PATH, SELFIE_PATH);
  const selfieData = fs.readFileSync(selfiePath);
  const selfieKey = `${randomUUID()}.jpg`;
  await blobStorage.upload(
    settings.SELFIES_BUCKET,
    selfieKey,
    selfieData,
    lookup(selfiePath) || "image/jpeg",
  );

  // Upload avatar image
  const avatarPath = path.join(ROOT_PATH, AVATAR_PATH);
  const avatarData = fs.readFileSync(avatarPath);
  const avatarKey = `${randomUUID()}.jpg`;
  await blobStorage.upload(
    settings.AVATARS_BUCKET,
    avatarKey,
    avatarData,
    lookup(avatarPath) || "image/jpeg",
  );

  // Check if user already exists
  const existing = await db.query.users.findFirst({
    where: eq(schema.users.auth0UserId, seedUserId),
  });

  if (existing) {
    console.info(
      `User with auth0_user_id '${settings.AUTH0_SEED_USER_ID}' already exists, skipping creation.`,
    );
  }

  let userId = existing?.id;

  // Only create user if they don't exist
  if (!existing) {
    const [user] = await db
      .insert(schema.users)
      .values({
        auth0UserId: seedUserId,
        selfieImageKey: selfieKey,
        avatarImageKey: avatarKey,
      })
      .returning();

    if (!user) throw new Error("Failed to create user");
    console.info(`Created user: ${user.id}`);
    userId = user.id;
  } else if (userId) {
    // Update existing user's images
    await db
      .update(schema.users)
      .set({ selfieImageKey: selfieKey, avatarImageKey: avatarKey })
      .where(eq(schema.users.id, userId));
    console.info(`Updated user ${userId} with new images`);
  }

  if (!userId) throw new Error("No user ID available");

  // Add wearables
  for (const [name, data] of Object.entries(WEARABLES)) {
    // Upload wearable image
    const imagePath = path.join(ROOT_PATH, data.imagePath);
    const imageData = fs.readFileSync(imagePath);
    const ext = path.extname(imagePath);
    const imageKey = `${randomUUID()}${ext}`;
    await blobStorage.upload(
      settings.WEARABLES_BUCKET,
      imageKey,
      imageData,
      lookup(imagePath) || "image/jpeg",
    );

    // Add wearable
    const [wearable] = await db
      .insert(schema.wearables)
      .values({
        userId,
        category: data.category,
        imageKey,
      })
      .returning();

    if (!wearable) throw new Error(`Failed to create wearable: ${name}`);

    // Upload WOA image
    const woaPath = path.join(ROOT_PATH, "images", "results", "human_4", "single", `${name}.jpg`);
    if (fs.existsSync(woaPath)) {
      const woaData = fs.readFileSync(woaPath);
      const woaKey = `${randomUUID()}.jpg`;
      await blobStorage.upload(settings.WOA_BUCKET, woaKey, woaData, "image/jpeg");

      // Upload mask image
      const maskPath = path.join(ROOT_PATH, "images", "masks", "human_4", "post", `${name}.jpg`);
      if (fs.existsSync(maskPath)) {
        const maskData = fs.readFileSync(maskPath);
        const maskKey = `${randomUUID()}.jpg`;
        await blobStorage.upload(settings.WOA_BUCKET, maskKey, maskData, "image/jpeg");

        // Add WearableOnAvatarImage
        await db.insert(schema.wearableOnAvatarImages).values({
          userId,
          avatarImageKey: avatarKey,
          wearableImageKey: imageKey,
          imageKey: woaKey,
          maskImageKey: maskKey,
        });
      }
    }

    console.info(`Added wearable: ${name} (${wearable.id})`);
  }

  console.info("Seeding complete!");
}

seed().catch((err) => {
  console.error("Seed failed:", err);
  process.exit(1);
});
